File: src/core/tool.ts
/**
 * Tool class for agentic tool-calling.
 *
 * Provides a type-safe way to define tools with Schema-based
 * parameter validation and optional result validation.
 *
 * @module core/tool
 */

import { Schema } from '../schema/index.js';
import type { JSONSchemaDocument, SchemaInstance } from '../schema/types.js';

/**
 * Tool execution context.
 */
export interface ToolContext {
  /** Current model being used */
  model: string;
  /** Current provider */
  provider: string;
  /** Abort signal for cancellation */
  signal?: AbortSignal;
}

/**
 * Tool execution input.
 */
export interface ToolInput<TArgs = unknown> {
  /** Parsed and validated arguments */
  args: TArgs;
  /** Execution context */
  ctx: ToolContext;
}

/**
 * Tool definition options.
 */
export interface ToolDefinition<TArgs = unknown, TResult = unknown> {
  /** Unique tool name (function name for the LLM) */
  name: string;

  /** Description of what the tool does */
  description: string;

  /** Parameter schema (Schema class instance or constructor) */
  parameters: SchemaInstance | (new () => SchemaInstance);

  /**
   * Execute function.
   * Receives validated args and context, returns result.
   */
  execute: (input: ToolInput<TArgs>) => Promise<TResult>;

  /**
   * Optional result schema for validation.
   */
  resultSchema?: SchemaInstance | (new () => SchemaInstance);

  /**
   * Timeout for this tool in milliseconds.
   * @default 10000
   */
  timeout?: number;
}

/**
 * Serialized tool call from the LLM.
 */
export interface ToolCall {
  /** Tool call ID (for response matching) */
  id: string;
  /** Tool name */
  name: string;
  /** Raw arguments (JSON-parsed) */
  args: unknown;
}

/**
 * Tool execution result.
 */
export interface ToolResult {
  /** Tool call ID */
  id: string;
  /** Result data or error */
  result: unknown;
  /** Whether execution succeeded */
  success: boolean;
  /** Error message if failed */
  error?: string;
}

/**
 * Tool class for defining callable tools.
 *
 * @example
 * ```typescript
 * class WeatherParams extends Schema {
 *   city = Schema.String({ description: "City name" });
 *   unit = Schema.Enum(["celsius", "fahrenheit"] as const, { optional: true });
 * }
 *
 * const weatherTool = new Tool({
 *   name: "get_weather",
 *   description: "Get current weather for a city",
 *   parameters: WeatherParams,
 *   execute: async ({ args }) => {
 *     const weather = await fetchWeather(args.city);
 *     return { temperature: weather.temp, conditions: weather.sky };
 *   },
 * });
 *
 * chat.addTool(weatherTool);
 * ```
 */
export class Tool<TArgs = unknown, TResult = unknown> {
  /** Tool name */
  readonly name: string;

  /** Tool description */
  readonly description: string;

  /** Parameter schema instance */
  readonly parameters: SchemaInstance;

  /** Result schema instance (optional) */
  readonly resultSchema?: SchemaInstance;

  /** Execution timeout */
  readonly timeout: number;

  /** Execute function */
  private readonly executeFn: (input: ToolInput<TArgs>) => Promise<TResult>;

  constructor(definition: ToolDefinition<TArgs, TResult>) {
    if (!definition.name || typeof definition.name !== 'string') {
      throw new TypeError('Tool name is required and must be a string');
    }
    if (!definition.description || typeof definition.description !== 'string') {
      throw new TypeError('Tool description is required and must be a string');
    }
    if (!definition.parameters) {
      throw new TypeError('Tool parameters schema is required');
    }
    if (typeof definition.execute !== 'function') {
      throw new TypeError('Tool execute function is required');
    }

    this.name = definition.name;
    this.description = definition.description;
    this.timeout = definition.timeout ?? 10_000;
    this.executeFn = definition.execute;

    // Instantiate schema if constructor passed
    this.parameters =
      typeof definition.parameters === 'function'
        ? Schema.create(definition.parameters)
        : definition.parameters;

    if (definition.resultSchema) {
      this.resultSchema =
        typeof definition.resultSchema === 'function'
          ? Schema.create(definition.resultSchema)
          : definition.resultSchema;
    }
  }

  /**
   * Get JSON Schema for parameters.
   */
  getParametersSchema(): JSONSchemaDocument {
    return Schema.toJSON(this.parameters);
  }

  /**
   * Validate arguments against parameter schema.
   */
  validateArgs(args: unknown): boolean {
    return Schema.validate(this.parameters, args);
  }

  /**
   * Parse and validate arguments.
   * Throws SchemaError if invalid.
   */
  parseArgs(args: unknown): TArgs {
    const instance = Schema.clone(this.parameters);
    Schema.parse(instance, args);
    return Schema.toObject(instance) as TArgs;
  }

  /**
   * Validate result against result schema (if defined).
   */
  validateResult(result: unknown): boolean {
    if (!this.resultSchema) return true;
    return Schema.validate(this.resultSchema, result);
  }

  /**
   * Execute the tool with validated arguments.
   */
  async execute(input: ToolInput<TArgs>): Promise<TResult> {
    return this.executeFn(input);
  }

  /**
   * Execute with timeout.
   */
  async executeWithTimeout(input: ToolInput<TArgs>): Promise<TResult> {
    return withTimeout(this.executeFn(input), this.timeout);
  }

  /**
   * Execute a tool call (validates args, executes, validates result).
   * Returns ToolResult with success/error info.
   */
  async executeCall(call: ToolCall, ctx: ToolContext): Promise<ToolResult> {
    try {
      // Validate arguments
      if (!this.validateArgs(call.args)) {
        return {
          id: call.id,
          result: null,
          success: false,
          error: `Invalid arguments for tool ${this.name}`,
        };
      }

      // Parse args
      const args = this.parseArgs(call.args);

      // Execute with timeout
      const result = await this.executeWithTimeout({ args, ctx });

      // Validate result
      if (!this.validateResult(result)) {
        return {
          id: call.id,
          result: null,
          success: false,
          error: `Invalid result from tool ${this.name}`,
        };
      }

      return {
        id: call.id,
        result,
        success: true,
      };
    } catch (error) {
      return {
        id: call.id,
        result: null,
        success: false,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }
}

/**
 * Helper to create a tool with type inference.
 */
export function createTool<TArgs, TResult>(
  definition: ToolDefinition<TArgs, TResult>,
): Tool<TArgs, TResult> {
  return new Tool(definition);
}

/**
 * Execute a promise with timeout.
 */
async function withTimeout<T>(promise: Promise<T>, ms: number): Promise<T> {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => reject(new Error('Tool execution timed out')), ms);

    promise
      .then((value) => {
        clearTimeout(timer);
        resolve(value);
      })
      .catch((error) => {
        clearTimeout(timer);
        reject(error);
      });
  });
}

File: ocaml/lib/core/tool.ml
type context = {
  model : string;
  provider : Chatoyant_provider.Provider.id;
}

type call = {
  id : string;
  name : string;
  arguments : Chatoyant_runtime.Json.t;
}

type result = {
  id : string;
  ok : bool;
  value : Chatoyant_runtime.Json.t option;
  error : string option;
}

type t = {
  tool_name : string;
  tool_description : string;
  tool_parameters : Chatoyant_schema.Schema.field;
  tool_result_schema : Chatoyant_schema.Schema.field option;
  execute :
    context -> Chatoyant_runtime.Json.t -> (Chatoyant_runtime.Json.t, string) Stdlib.result;
}

let create ~name ~description ~parameters ?result_schema execute =
  { tool_name = name; tool_description = description; tool_parameters = parameters; tool_result_schema = result_schema; execute }

let create_typed ~name ~description ~args ~result execute =
  create ~name ~description ~parameters:(Chatoyant_schema.Codec.schema args)
    ~result_schema:(Chatoyant_schema.Codec.schema result)
    (fun context json ->
      match Chatoyant_schema.Codec.decode args json with
      | Error message -> Error message
      | Ok args -> (
          match execute context args with
          | Error _ as err -> err
          | Ok value -> Ok (Chatoyant_schema.Codec.encode result value)))

let name tool = tool.tool_name
let description tool = tool.tool_description
let parameters tool = tool.tool_parameters
let json_schema tool = Chatoyant_schema.Schema.to_json_schema tool.tool_parameters

let to_provider_definition (tool : t) : Chatoyant_provider.Provider.tool_definition =
  {
    Chatoyant_provider.Provider.tool_name = tool.tool_name;
    tool_description = Some tool.tool_description;
    tool_parameters = json_schema tool;
    tool_strict = Some true;
  }

let failure id message = { id; ok = false; value = None; error = Some message }
let success id value = { id; ok = true; value = Some value; error = None }

let execute_call context call tool =
  if call.name <> tool.tool_name then failure call.id ("Unknown tool: " ^ call.name)
  else
    match Chatoyant_schema.Value.validate tool.tool_parameters call.arguments with
    | Error error -> failure call.id (Chatoyant_schema.Value.error_to_string error)
    | Ok () -> (
        match tool.execute context call.arguments with
        | Error message -> failure call.id message
        | Ok value -> (
            match tool.tool_result_schema with
            | None -> success call.id value
            | Some schema -> (
                match Chatoyant_schema.Value.validate schema value with
                | Ok () -> success call.id value
                | Error error -> failure call.id (Chatoyant_schema.Value.error_to_string error))))

let string value = Chatoyant_runtime.Json.String value

let add_opt name value fields =
  match value with
  | None -> fields
  | Some value -> (name, value) :: fields

let call_to_json (call : call) =
  Chatoyant_runtime.Json.Object
    [
      ("id", string call.id);
      ("name", string call.name);
      ("arguments", call.arguments);
    ]

let result_to_json (result : result) =
  [
    ("id", string result.id);
    ("ok", Chatoyant_runtime.Json.Bool result.ok);
  ]
  |> add_opt "value" result.value
  |> add_opt "error" (Option.map string result.error)
  |> List.rev |> fun fields -> Chatoyant_runtime.Json.Object fields

module Args = struct
  let ( let* ) result fn = Stdlib.Result.bind result fn

  let field name json =
    match Chatoyant_runtime.Json.field name json with
    | Some value -> Ok value
    | None -> Error ("missing field: " ^ name)

  let type_error name expected =
    Error ("expected field " ^ name ^ " to be " ^ expected)

  let string name json =
    let* value = field name json in
    match Chatoyant_runtime.Json.as_string value with
    | Some value -> Ok value
    | None -> type_error name "a string"

  let float name json =
    let* value = field name json in
    match Chatoyant_runtime.Json.as_float value with
    | Some value -> Ok value
    | None -> type_error name "a number"

  let int name json =
    let* value = field name json in
    match Chatoyant_runtime.Json.as_int value with
    | Some value -> Ok value
    | None -> type_error name "an integer"

  let bool name json =
    let* value = field name json in
    match Chatoyant_runtime.Json.as_bool value with
    | Some value -> Ok value
    | None -> type_error name "a boolean"

  let object_ name json =
    let* value = field name json in
    match Chatoyant_runtime.Json.as_object value with
    | Some value -> Ok value
    | None -> type_error name "an object"

  let list name json =
    let* value = field name json in
    match Chatoyant_runtime.Json.as_list value with
    | Some value -> Ok value
    | None -> type_error name "an array"

  let optional decode name json =
    match Chatoyant_runtime.Json.field name json with
    | None | Some Chatoyant_runtime.Json.Null -> Ok None
    | Some _ -> Stdlib.Result.map Option.some (decode name json)
end

module Json = struct
  let null = Chatoyant_runtime.Json.Null
  let string value = Chatoyant_runtime.Json.String value
  let float value = Chatoyant_runtime.Json.Float value
  let int value = Chatoyant_runtime.Json.Float (Float.of_int value)
  let bool value = Chatoyant_runtime.Json.Bool value
  let object_ fields = Chatoyant_runtime.Json.Object fields
  let array values = Chatoyant_runtime.Json.Array values
end
